### PR TITLE
Update for wildcard entries in hosts file param

### PR DESCRIPTION
**Pull request by titpetric**
Added support for wildcard entries in the hosts file, using \* as a prefix:

```
192.168.0.1   *.example.com
```

Only one wildcard can be used, and it must be on the beginning of the domain. Name resolution is implemented on a single level, so the example will resolve `api.example.com`, but will not resolve `v2.api.example.com`.

### CODE REVIEW

File: hostsfile/utils.go [hunk 1]
```diff
@@ -36,8 +41,56 @@ func newHostlist(data []byte) *hostlist {
...
+					var left string;
+					left = name[0:len(name)-len(hostname.domain)]
+					if !strings.Contains(left, ".") {
+						addrs = append(addrs, hostname.ip)
+					}
+				}
+			}
+		}
+	}
+
```
Comment (janeczku) [inline hostsfile/utils.go:90]: I am not a fan of the ~~trailing~~ leading dot in the domain value for wildcard hostnames. They are already identified by the `wildcard` field. Here is what i suggest instead:

``` Go
    var parent string
    i := strings.IndexByte(name, '.')
    if i > -1 && i < len(name)-1 {
        parent = name[i+1:]
    }

    if len(addrs) == 0 && len(parent) > 0 {
        for _, hostname := range *h {
            if hostname.wildcard == true && hostname.domain == parent {
                addrs = append(addrs, hostname.ip)
            }
        }
    }
```

Also thanks to you we have a properly working `hostname.Equal()` method now, so we should probably leverage that to check for exact duplicates in the `(h *hostlist) add(hostnamev *hostname)` func. See https://github.com/titpetric/go-dnsmasq/blob/master/hostsfile/utils.go#L95.

File: hostsfile/utils.go [hunk 2]
```diff
@@ -114,8 +167,14 @@ func parseLine(line string) hostlist {
 		return hostnames
 	}
 
+	var isWildcard bool
 	for _, v := range domains {
-		hostname := newHostname(v, ip, isIPv6)
+		isWildcard = false
+		if v[0:1] == "*" {
+			v = v[1:]
```
Comment (janeczku) [inline hostsfile/utils.go]:
``` Go
        if v[0:2] == "*." {
            v = v[2:]
```

See my other comment.

File: hostsfile/hostsfile_test.go [hunk 3]
```diff
@@ -74,30 +105,78 @@ func TestParseLine(t *testing.T) {
...
+
+	// IPv6 (not link-local)
+	hosts = parseLine("2a02:7a8:1:250::80:1		rtvslo.si img.rtvslo.si")
+	if !hosts.Contains(newHostname("img.rtvslo.si", net.ParseIP("2a02:7a8:1:250::80:1"), true, false)) ||
+		len(hosts) != 2 {
+		t.Error("Expected to find rtvslo.si ipv6, two hosts")
+	}
+
+	/* the following all fails since the addressses are link-local */
+
```
Comment (janeczku) [inline hostsfile/hostsfile_test.go:171]: Yeah....the loopback addresses are not parsed because they are matched by the `net.IsGlobalUnicast()` func. https://golang.org/src/net/ip.go?s=4290:4325#L146
We should probably allow them as it is technically legal to return them in a DNS answer. But this will be another PR.

**Comment by janeczku** [comment]
Thanks for the PR and especially for fixing and adding the tests! :thumbsup: :smiley:

File: hostsfile/utils.go (hunk 1 above)
Comment (titpetric) [inline hostsfile/utils.go:90]: I used a slightly different concat strategy (strings.Join). Also fixed add() with using host.Equal + added tests to check that error is thrown where it should be.

File: hostsfile/utils.go (hunk 2 above)
Comment (titpetric) [inline hostsfile/utils.go]: ACK.

File: hostsfile/hostsfile_test.go (hunk 3 above)
Comment (titpetric) [inline hostsfile/hostsfile_test.go:171]: Enjoy that one, it's out of my scope :) :+1:

**Comment by janeczku** [comment]
Could you squash the commits?

**Comment by titpetric** [comment]
If you insist, I can, but to be honest, I don't want to. I'd rather send you a postcard from the Azores, while I'm enjoying some wine with my lovely wife. Merge it and let's continue. I'd rather make some usable feature than concern myself with commit history.

**Comment by janeczku** [comment]
Don't forget the postcard!

**Comment by titpetric** [comment]
Need an address for that one ;)
